config/configmap.go

```diff
@@ -170,51 +163,62 @@ var (
+		// ensure we are dealing with map to map comparison
+		if from.Kind() == reflect.Map && to.Kind() == reflect.Map {
+			toElem := to.Type().Elem()
+			// ensure that map values are pointers to a struct
+			// (that may be nil and require manual setting w/ zero value)
+			if toElem.Kind() == reflect.Ptr && toElem.Elem().Kind() == reflect.Struct {
+				fromRange := from.MapRange()
+				for fromRange.Next() {
+					fromKey := fromRange.Key()
+					fromValue := fromRange.Value()
+					// ensure that we've run into a nil pointer instance
+					if fromValue.IsNil() {
+						newFromValue := reflect.New(toElem.Elem())
+						from.SetMapIndex(fromKey, newFromValue)
+					}
 				}
 			}
 		}
+		return from.Interface(), nil
 	}
-	return from.Interface(), nil
 }
 
-// mapStringToMapComponentIDHookFunc returns a DecodeHookFunc that converts a map[string]interface{} to
-// map[ComponentID]interface{}.
-// This is needed in combination since the ComponentID.UnmarshalText may produce equal IDs for different strings,
+// mapKeyStringToMapKeyTextUnmarshalerHookFunc returns a DecodeHookFuncType that checks that a conversion from
+// map[string]interface{} to map[encoding.TextUnmarshaler]interface{} does not overwrite keys,
+// when UnmarshalText produces equal elements from different strings (e.g. trims whitespaces).
+//
+// This is needed in combination with ComponentID, which may produce equal IDs for different strings,
 // and an error needs to be returned in that case, otherwise the last equivalent ID overwrites the previous one.
-var mapStringToMapComponentIDHookFunc = func(f reflect.Type, t reflect.Type, data interface{}) (interface{}, error) {
-	if f.Kind() != reflect.Map || f.Key().Kind() != reflect.String {
-		return data, nil
-	}
+func mapKeyStringToMapKeyTextUnmarshalerHookFunc() mapstructure.DecodeHookFuncType {
+	return func(f reflect.Type, t reflect.Type, data interface{}) (interface{}, error) {
+		if f.Kind() != reflect.Map || f.Key().Kind() != reflect.String {
+			return data, nil
+		}
 
-	if t.Kind() != reflect.Map || t.Key() != componentIDType {
-		return data, nil
-	}
+		if t.Kind() != reflect.Map {
+			return data, nil
+		}
 
-	m := make(map[ComponentID]interface{})
-	for k, v := range data.(map[string]interface{}) {
-		id, err := NewComponentIDFromString(k)
-		if err != nil {
-			return nil, err
+		if _, ok := reflect.New(t.Key()).Interface().(encoding.TextUnmarshaler); !ok {
+			return data, nil
 		}
-		if _, ok := m[id]; ok {
-			return nil, fmt.Errorf("duplicate name %q after trimming spaces %v", k, id)
+
+		m := reflect.MakeMap(reflect.MapOf(t.Key(), reflect.TypeOf(true)))
+		for k := range data.(map[string]interface{}) {
+			tKey := reflect.New(t.Key())
+			if err := tKey.Interface().(encoding.TextUnmarshaler).UnmarshalText([]byte(k)); err != nil {
+				return nil, err
+			}
+
+			if m.MapIndex(reflect.Indirect(tKey)).IsValid() {
+				return nil, fmt.Errorf("duplicate name %q after trimming spaces %v", k, tKey)
```

Comment:
This error may be true in one particular case, but it's not with a general `UnmarshalText` implementation
```suggestion
				return nil, fmt.Errorf("duplicate name %q after unmarshaling %v", k, tKey)
```